src/OpenTelemetry.Exporter.Geneva/GenevaTableName.cs

```diff
@@ -0,0 +1,51 @@
+// <copyright file="GenevaTableName.cs" company="OpenTelemetry Authors">
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Text;
+
+namespace OpenTelemetry.Exporter.Geneva;
+
+/// <summary>
+/// Stores details about a Geneva table name.
+/// </summary>
+public readonly ref struct GenevaTableName
```

Comment:
Could we get rid of this struct? I think we could provide a more user-friendly API to achieve this with just the `IGenevaTableColumnNameResolver`. How about this?

1. We update the public API to provide the string representation of the mapped table name:

```diff
namespace OpenTelemetry.Exporter.Geneva
{
   public class GenevaExporterOptions
   {
+      public IGenevaTableColumnNameResolver TableColumnNameResolver { get; set; }
   }

+   public interface IGenevaTableColumnNameResolver
+   {
+       ISet<string> ResolveColumnNamesForTableName(string tableName);
+   }
```

2. Update `TableNameCacheDictionary` in TableNameSerializer to use:
```csharp
// This would allow us to store the string representation of the mapped table name.
// We would not be allocating a new string on hot path
Dictionary<string, Tuple<string, byte>>()
```

3. Update the method signature of `ResolveAndSerializeTableNameForCategoryName` to: 
```csharp
ResolveAndSerializeTableNameForCategoryName(byte[] destination, int offset, string categoryName, Tuple<string, byte[]> mappedTableNameData)
```
4. Update `Serialize` method in MsgPackLogExporter to:
```csharp
...
// Part A - core envelope
cursor = AddPartAField(buffer, cursor, Schema.V40.PartA.Name, eventName.Item2);

...
var columnNames = this.m_tableColumnNameResolver?.ResolveColumnNamesForTableName(eventName.Item1)
            ?? this.m_defaultColumnNames;
```

5. Sample TableColumnNameResolver:
```csharp
    internal sealed class TableColumnNameResolver : IGenevaTableColumnNameResolver
    {
        public const string TestTableName = "TestTable";

        private readonly ISet<string> testColumnNames = new HashSet<string>
        {
            "food",
            "price",
        };

        public ISet<string> ResolveColumnNamesForTableName(string tableName)
        {
            if (string.Equals(TestTableName, tableName, StringComparison.Ordinal)
            {
                return this.testColumnNames;
            }

            return null;
        }
    }
```

Reply:
There are cases where we _never_ have the table name as a string. We sanitize category name directly onto the output buffer as bytes. So if we made this API string-based, we would potentially be forcing string allocations onto users who do not care about this feature 😢 We could just pass `ReadOnlySpan<byte> tableName` instead of the struct but that is worse IMO because it really paints us into a corner as far as ever changing the implementation.

Reply:
> There are cases where we never have the table name as a string. We sanitize category name directly onto the output buffer as bytes

I guess this is another reason why caching the mapped name for every category would make things easy for us. (Referring to this conversation: https://github.com/open-telemetry/opentelemetry-dotnet-contrib/pull/796#discussion_r1036641979) 😀

Since we are going to do a dictionary lookup for each cateogry anyway, why not just store the mapped table name's `string` and `byte[]` representation for each category. I don't think storing an extra `string` along with the `byte[]` should be a matter of concern as these are one-time allocations (only when adding a new entry to cache).

 If storing the extra string in the cache still seems like a penalty for users who won't use the `IGenevaTableColumnNameResolver` then we could have two kinds of caches based on the user's provided options: one with only `byte[]` (if they don't use this feature) and one with both `string` and `byte[]` (if they use this feature)

I think a string-based API for this feature is more intuitive and user friendly and if some additional memory usage to also cache mapped table name strings helps us achieve that, we should consider it.

Reply:
> I think a string-based API for this feature is more intuitive and user friendly

100% agree. Although intuitive and user friendly API wasn't necessarily a goal 😄 A nice-to-have, certainly.

> Since we are going to do a dictionary lookup for each cateogry anyway, why not just store the mapped table name's string and byte[] representation for each category. I don't think storing an extra string along with the byte[] should be a matter of concern as these are one-time allocations (only when adding a new entry to cache).

Ya I remember the discussion. Basically do we want an unconstrained cache or not. It still worries me, but since you keep asking for it I went ahead and made the change 😄

> we could have two kinds of caches based on the user's provided options

Doesn't feel worth the added complexity.